rfd/0abc-kube-access-forward-identity.md

```diff
@@ -0,0 +1,205 @@
+the time it takes to forward a new request upstream. Currently, the proxy creates a new
+`http.Transport` per request which means that a new connection is established
+every time a request is forwarded. This is not ideal because it increases the
+latency of the request and it also increases the load on the upstream service.
+
+This option requires sending the user's IP address to the upstream service
+using a header instead of the protocol extension implemented by [TLS IP Pinning RFD](https://github.com/gravitational/teleport/pull/22481).
+This is because the connection is reused and the IP address is not available
+in the connection prelude since the connection does not belong to a specific user/request.
+
+## Rollout plan
+
+To avoid breaking changes, we will implement this feature in two phases.
+
+### Phase 1: Add support for receiving user identity in Kube Service/Proxy from the Kube Proxy
+
+- Teleport 13.0
+
+We will add support for receiving user identity in Kube Service/Proxy from the
+Kube Proxy. Teleport heartbeats contain the agent version and the proxy will
+only forward the user's identity to the Kube Service/Proxy if the agent version
+is >= 13.0. For older agents, the proxy will continue to generate a new certificate
+for each user's identity and call the `ProcessKubeCSR` endpoint. This will allow
+us to roll out this feature without breaking changes to Teleport clusters running
+older versions (<= 12.0).
+
+When the proxy uses its X.509 certificate to authenticate to the upstream service,
+the upstream service will expect the user's identity to be
+forwarded using HTTP headers or Proxy Protocol extensions. If the user's identity
+is not forwarded, the upstream service will reject the request.
+
+### Phase 2: Enable the proxy forwarding of user identity to Kube Service/Proxy
+
+- Teleport 14.0
+
+We will enable the proxy to forward user identity to Kube Service/Proxy.
+At this point, Teleport supports clients running Teleport 13.0 and newer and we can
+remove the code that generates a new certificate for each user's identity and calls
+`ProcessKubeCSR` endpoint because the upstream services already support receiving
+user identity from the proxy without the need for the proxy to generate a new certificate.
+
+## Enterprise considerations
+
+Currently, when the Teleport proxy calls `ProcessKubeCSR` endpoint and Auth server
+is configured with an Enterprise license, the auth server checks if the license
+has the Kubernetes feature enabled. If the Auth server isn't licensed to Kubernetes usage,
+Auth server will return an error to the proxy and the request is rejected.
+
+Since we no longer need to call `ProcessKubeCSR` endpoint, we will need to verify
+that the cluster is licensed to Kubernetes before forwarding the request to the Kube
+Service. This can be done by adding a new endpoint to the Auth server that will
+return the license information. The proxy will call this endpoint periodically and
```

Comment:
Auth server's ping endpoint already returns this information:

https://github.com/gravitational/teleport/blob/afcac88b2859409a4ac2bd3264193932a164cbb7/api/proto/teleport/legacy/client/proto/authservice.proto#L540

Reply:
Rewrote the RFD wording and included a stronger check so we don't end up relying only on ping.

Auth will validate the license when GenUserCerts is invoked by the user and it includes the `KubernetesCluster` information:

> Since we no longer need to call `ProcessKubeCSR` endpoint, we will need to verify
that the cluster is licensed to Kubernetes before forwarding the request to the Kube
Service. This can be done in `GenerateUserCerts` each time a user requests a
new certificate signing request for a Kubernetes cluster. Auth server will check if the
license has the Kubernetes feature enabled and if it doesn't, it will return an error
to the user blocking him from using Kubernetes Access. Proxy and Kube Service will
also check if the license has the Kubernetes feature enabled and if it doesn't,
they will return an error to the user. Services will call Auth server's `Ping`
endpoint to check if the license flags include the Kubernetes feature. The proxy
will call this endpoint periodically and if the license doesn't have the Kubernetes
feature enabled, the proxy will reject the request. This would allow us to
automatically fix the error message returned to the user when the license is
invalid [teleport.e#661](https://github.com/gravitational/teleport.e/issues/661).

Reply:
> Sounds fine in general but tbh I'm not sure I understand what the additional check in GenerateUserCerts buys us?
> 
> Main thing is to make run the check is done on the agent side as well to account for trusted clusters scenario (i.e. root is licensed for Kube, leaf is not - root's auth check won't help here).

@r0mant 

My main point was to ensure that auth validates it because it's quite easy to remove the ping from kube_proxy/agent and deploy the custom versions without that verification.  Since Kube access is open source, users won't lose any Kube feature and we could not ensure that auth is licensed.

Now that you mentioned it, I think the easiest and simplest way is just to deny the `UpsertKubeServer` if the auth server is not licensed to Kubernetes. This will prevent Kube access to work and we have a strong validation at the root and leaf clusters.

This changes the heartbeat behavior since Kube is not allowed to register the `kube_server` but we can print a detailed log message